source/include/azure_iot.h

```diff
@@ -0,0 +1,153 @@
+                                                 uint32_t ulBufferLength );
+
+/**
+ * @brief Append a property name and value to a message.
+ *
+ * @param[in] pxMessageProperties The #AzureIoTMessageProperties_t* to use for the operation.
+ * @param[in] pucName The name of the property to append.
+ * @param[in] ulNameLength The length of \p pucName.
+ * @param[in] pucValue The value of the property to append.
+ * @param[in] ulValueLength The length of \p pucValue.
+ * @return An #AzureIoTResult_t with the result of the operation.
+ */
+AzureIoTResult_t AzureIoT_MessagePropertiesAppend( AzureIoTMessageProperties_t * pxMessageProperties,
+                                                   const uint8_t * pucName,
+                                                   uint32_t ulNameLength,
+                                                   const uint8_t * pucValue,
+                                                   uint32_t ulValueLength );
+
+/**
+ * @brief Find a property in the message property bag.
+ *
+ * @param[in] pxMessageProperties The #AzureIoTMessageProperties_t* to use for the operation.
+ * @param[in] pucName The name of the property to find.
+ * @param[in] nameLength Length of the property name.
+ * @param[out] ppucOutValue The output pointer to the property value.
+ * @param[out] pulOutValueLength The length of \p ppucOutValue.
+ * @return An #AzureIoTResult_t with the result of the operation.
+ */
+AzureIoTResult_t AzureIoT_MessagePropertiesFind( AzureIoTMessageProperties_t * pxMessageProperties,
+                                                 const uint8_t * pucName,
+                                                 uint32_t ulNameLength,
+                                                 const uint8_t ** ppucOutValue,
+                                                 uint32_t * pulOutValueLength );
+
+/**
+ * @brief As part of symmetric key authentication, HMAC256 a buffer of bytes and base64 encode the result.
+ *
+ * @note This is used within Azure IoT Hub and Device Provisioning APIs if a symmetric key is set.
+ *
+ * @param[in] xAzureIoTHMACFunction The #AzureIoTGetHMACFunc_t function to use for HMAC256 hashing.
+ * @param[in] pucKey A pointer to the base64 encoded key.
+ * @param[in] ulKeySize The length of the \p pucKey.
+ * @param[in] pucMessage A pointer to the blob to be hashed.
+ * @param[in] ulMessageSize The length of \p pucMessage.
+ * @param[in] pucBuffer An intermediary buffer to put the base64 decoded key.
+ * @param[in] ulBufferLength The length of \p pucBuffer.
+ * @param[out] pucOutput The buffer into which the resulting HMAC256 hashed, base64 encoded message will
+ * be placed.
+ * @param[in] ulOutputSize Size of \p pucOutput.
+ * @param[out] pulOutputLength The output length of \p pucOutput.
+ * @return An #AzureIoTResult_t with the result of the operation.
+ */
+AzureIoTResult_t AzureIoT_Base64HMACCalculate( AzureIoTGetHMACFunc_t xAzureIoTHMACFunction,
```

Comment:
Is this a function we think applications would ever call directly?  If not could we move it to a private header file?

Reply:
And assuming this can move to a private header, related struct/function pointers (e.g. AzureIoTGetHMACFunc_t) could move as well??

Reply:
I say put it in private and we can always expose if needed. We can't go the other way.

Reply:
I do want to do that in another PR though. There's a greater discussion that needs to happen around directory structure and what to name things. I'll make a note to tackle.